elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +265,144 @@ private Object fetchRelationship(Environment context) {
+            } else {
+                return updateObject(); //TODO: implement update
+            }
+        } catch (BadRequestException e) {
+            log.debug(e.toString());
+            return createObject(input, outputType, requestScope, Optional.of(ids), field, parentSource);
+        }
+    }
+
+    /**
+     * creates or updates a new/existing entity
+     * @param idFieldName field name
+     * @param input input data object
+     * @param requestScope request scope
+     * @param outputType graphql-java output type
+     * @return list/set of persistent resource objects
+     */
+    private PersistentResource upsertObject(String idFieldName, Map<String, Object> input,
+                                RequestScope requestScope,  Field field, GraphQLType outputType) {
+        /* fetch id values fed to 'data' argument */
+        Set<String> ids = input.entrySet().stream().filter(entry -> idFieldName.equalsIgnoreCase(entry.getKey()))
+                .map(entry -> (String) entry.getValue())
+                .collect(Collectors.toSet());
+        try {
+            Set<PersistentResource> fetchEntries = (Set) fetchObject(requestScope, outputType,
+                    Optional.of(new ArrayList<>(ids)), Optional.empty(),
+                    Optional.empty(), Optional.empty(), Optional.empty());
+            if(fetchEntries.isEmpty()) { /* empty set returned, must create new */
+                return createObject(input, outputType, requestScope, Optional.of(ids), field, null);
+            } else { /* must update object */
+                return updateObject(); //TODO: implement update
+            }
+        } catch(BadRequestException e) { /* no id(s) provided, must create new */
+            log.debug(e.toString());
+            return createObject(input, outputType, requestScope, Optional.empty(), field, null);
+        }
+    }
+
+    /**
+     * Creates a new "root-level" resource in the database
+     * @param input individual input data objects
+     * @param outputType graphql-java output type
+     * @param requestScope request scope
+     * @param ids id
+     * @return persistent resource object
+     */
+    private PersistentResource createObject(Map<String, Object> input, GraphQLType outputType,
+                                RequestScope requestScope, Optional<Set<String>> ids,
+                                Field field, PersistentResource parent) {
+        GraphQLObjectType objectType = (GraphQLObjectType) ((GraphQLList) outputType).getWrappedType();
+        EntityDictionary dictionary = requestScope.getDictionary();
+
+        Optional<String> uuid = ids.map(strings -> strings.iterator().next());
+        Class<?> entityClass;
+        PersistentResource toCreate;
+
+        if(parent == null) {
+            entityClass = dictionary.getEntityClass(objectType.getName());
+            toCreate = PersistentResource.createObject(null, entityClass, requestScope, uuid);
```

Comment:
There is a function in PersistentResource that doesn't require a parent.  You can just remove the first null param.